pyro/optim/multi.py

```diff
@@ -0,0 +1,114 @@
+        ``loss`` tensor.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        """
+        raise NotImplementedError
+
+
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        names = []
+        values = []
+        for name, value in params.items():
+            names.append(name)
+            values.append(value)
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
+        self.optim(values)
+
+
+class TorchMultiOptimizer(PyroMultiOptimizer):
+    """
+    Facade to wrap :class:`~torch.optim.Optimizer` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim_constructor, optim_args):
+        optim = PyroOptim(optim_constructor, optim_args)
+        super(TorchMultiOptimizer, self).__init__(optim)
+
+
+class MixedMultiOptimizer(MultiOptimizer):
+    """
+    Container class to combine different :class:`MultiOptimizer` instances for
+    different parameters.
+
+    :param list parts: A list of ``(names, optim)`` pairs, where each
+        ``names`` is a list of parameter names, and each ``optim`` is a
+        :class:`MultiOptimizer` object to be used for the named parameters.
+        Together the ``names`` should partition up all desired parameters to
+        optimize.
+    """
+    def __init__(self, parts):
+        self.parts = list(parts)
+
+    def step(self, loss, params):
+        for names_part, optim in self.parts:
+            optim.step(loss, {name: params[name] for name in names_part})
+
+
+class Newton2d(MultiOptimizer):
+    """
+    Implementation of :class:`MultiOptimizer` that performs a Newton update
+    on batched 2d variables, optionally regularizing via a per-parameter
+    ``trust_radius``. See :func:`~pyro.ops.newton.newton_step_2d` for details.
+
+    :param dict trust_radii: a dict mapping parameter name to radius of trust
+        region. Missing names will use unregularized Newton update, equivalent
+        to infinite trust radius.
+    """
+    def __init__(self, trust_radii={}):
+        self.trust_radii = trust_radii
+
+    def step(self, loss, params):
+        updated_values = {}
+        for name, value in params.items():
+            trust_radius = self.trust_radii.get(name)
+            updated_value, cov = newton_step_2d(loss, value, trust_radius)
+            updated_values[name] = updated_value.detach()
+        for name, value in params.items():
+            with torch.no_grad():
```

Comment:
We'll need to avoid the `no_grad` to support differentiating through the optimization process, but I believe that will require avoiding in-place tensor update and instead using `ParamStore.replace_param()`. I'm not sure the cleanest way to do this.

Reply:
Seems like we should just make it easier to update the param store as discussed yesterday, perhaps by adding a `force` flag or `set_tensor` kwarg to `param(...)`.

Reply:
This PR is fine as-is but we should do that next